lib/rspec/mocks/targets.rb

```diff
@@ -38,7 +38,9 @@ def self.disallow_negation(method_name)
     private
 
       def matcher_allowed?(matcher)
-        matcher.class.name.start_with?("RSpec::Mocks::Matchers".freeze)
+        ancestors = matcher.class.ancestors
+        namespace = "RSpec::Mocks::Matchers"
+        ancestors.any? { |klass| klass.name.to_s.start_with? namespace }
```

Comment:
> It seems offhand, that if we go this way, it can lead us to extraction of the "essential/mockable" part of the matcher to the module. For now this is just a speculation, I need time to play with the idea.

No. Delegation is more useful than inheritance in this fashion. We don't want to encourage more inheritance of our private APIs only support reasonable use cases like this. This module should only be used to improve the detection of inherited matchers from our own.